File: destack-py/destack/language/universe/universe.py
from typing import TYPE_CHECKING, Optional

from destack.language import Node, NodeType, builtin_node, builtin_property_parent

if TYPE_CHECKING:
    from destack.language import Space


@builtin_node(NodeType.UNIVERSE, is_abstract=True)
class Universe(Node):
    """
    The Universe of Destack.
    An abstract container for useful constants.
    """

    parent: Optional["Space"] = builtin_property_parent()

File: destack-ts/src/language/universe/universe.ts
import type { NodeReference } from "@destack/language/core";
import { Entity, Node, NodeType } from "@destack/language/core";
import { registerNodeClass } from "@destack/language/registry";
import type { Space } from "@destack/language/universe/space";

/* ==== DESTACK_GENERATED_START:NODE:20000 ==== */
/**
 * The Universe of Destack.
 * An abstract container for useful constants.
 */
export abstract class Universe extends Node {
  static metatype: NodeType = NodeType.UNIVERSE;

  /**
   * Node.parent
   */
  abstract get parent(): Entity | null;
  declare readonly parentPtr: NodeReference | null;

  /**
   * The Space this Node is in.
   */
  abstract get space(): Space | null;
  declare readonly spacePtr: NodeReference;

  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}
registerNodeClass(NodeType.UNIVERSE, Universe);
/* ==== DESTACK_GENERATED_END:NODE:20000 ==== */
